plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenPath.java

```diff
@@ -0,0 +1,377 @@
+  }
+
+  @Nullable
+  public Coordinates getCoordinates() { return coordinates; }
+
+  /**
+   * Returns {@code true} if this path is subordinate of another path.
+   *
+   * @see {@link #subordinateOf()}
+   */
+  public boolean isSubordinate() {
+    return isHash() || isSignature();
+  }
+
+  /**
+   * Returns {@code true} if this path represents a hash.
+   */
+  public boolean isHash() {
+    return hashType != null;
+  }
+
+  /**
+   * Returns {@code true} if this path represents a signature.
+   */
+  public boolean isSignature() {
+    return getCoordinates() != null && getCoordinates().getSignatureType() != null;
+  }
+
+  /**
+   * Returns the "main", non-subordinate path of this path. The "main" path is never a hash.
+   */
+  @Nonnull
+  public MavenPath main() {
+    MavenPath mavenPath = this;
+    while (mavenPath.isSubordinate()) {
+      mavenPath = mavenPath.subordinateOf();
+    }
+    return mavenPath;
+  }
+
+  /**
+   * Returns the "parent" path, that this path is subordinate of, or {@code null}.
+   */
+  @Nullable
+  public MavenPath subordinateOf() {
+    Coordinates mainCoordinates = null;
+    if (isHash()) {
+      int hashSuffixLen = hashType.getExt().length() + 1; // the dot
+      if (getCoordinates() != null) {
+        mainCoordinates = new Coordinates(
+            getCoordinates().isSnapshot(),
+            getCoordinates().getGroupId(),
+            getCoordinates().getArtifactId(),
+            getCoordinates().getVersion(),
+            getCoordinates().getBaseVersion(),
+            getCoordinates().getClassifier(),
+            getCoordinates().getExtension().substring(0, getCoordinates().getExtension().length() - hashSuffixLen),
+            getCoordinates().getSignatureType()
+        );
+      }
+      return new MavenPath(
+          path.substring(0, path.length() - hashSuffixLen),
+          mainCoordinates
+      );
+    }
+    else if (isSignature()) { // coordinates != null && coordinates.signatureType != null
+      int signatureSuffixLen = getCoordinates().getSignatureType().getExt().length() + 1; // the dot
+      mainCoordinates = new Coordinates(
+          getCoordinates().isSnapshot(),
+          getCoordinates().getGroupId(),
+          getCoordinates().getArtifactId(),
+          getCoordinates().getVersion(),
+          getCoordinates().getBaseVersion(),
+          getCoordinates().getClassifier(),
+          getCoordinates().getExtension().substring(0, getCoordinates().getExtension().length() - signatureSuffixLen),
+          null
+      );
+      return new MavenPath(
+          path.substring(0, path.length() - signatureSuffixLen),
+          mainCoordinates
+      );
+    }
+    return null;
+  }
+
+  /**
+   * Returns path of passed in hash type that is subordinate of this path. This path cannot be hash.
+   */
+  @Nonnull
+  public MavenPath hash(final HashType hashType) {
+    checkNotNull(hashType);
+    checkArgument(!isHash(), "This path is already a hash: " + this);
```

Comment:
do not use + syntax for precodnitions, use %s (goes for all uses you have for this in this PR)